subsys/net/lib/http/http_app_server.c

```diff
@@ -0,0 +1,1085 @@
+	size_t recv_len;
+	size_t pkt_len;
+
+	recv_len = net_pkt_appdatalen(pkt);
+	if (recv_len == 0) {
+		/* don't print info about zero-length app data buffers */
+		goto quit;
+	}
+
+	if (status) {
+		NET_DBG("[%p] Status %d <%s>", ctx, status, RC_STR(status));
+		goto out;
+	}
+
+	/* Get rid of possible IP headers in the first fragment. */
+	frag = pkt->frags;
+
+	pkt_len = net_pkt_get_len(pkt);
+
+	if (recv_len < pkt_len) {
+		net_buf_pull(frag, pkt_len - recv_len);
+		net_pkt_set_appdata(pkt, frag->data);
+	}
+
+	NET_DBG("[%p] Received %zd bytes http data", ctx, recv_len);
+
+	if (ctx->state == HTTP_STATE_OPEN) {
+		/* We have active websocket session and there is no longer
+		 * any HTTP traffic in the connection. Give the data to
+		 * application to send.
+		 */
+		goto http_only;
+	}
+
+	while (frag) {
+		/* If this fragment cannot be copied to result buf,
+		 * then parse what we have which will cause the callback to be
+		 * called in function on_body(), and continue copying.
+		 */
+		if ((ctx->http.data_len + frag->len) >
+		    ctx->request_buf_len) {
+
+			if (ctx->state == HTTP_STATE_HEADER_RECEIVED) {
+				goto http_ready;
+			}
+
+			/* If the caller has not supplied a callback, then
+			 * we cannot really continue if the request buffer
+			 * overflows. Set the data_len to mark how many bytes
+			 * should be needed in the response_buf.
+			 */
+			if (http_process_recv(ctx) < 0) {
+				ctx->http.data_len = recv_len;
+				goto out;
+			}
+
+			parsed_len =
+				http_parser_execute(&ctx->http.parser,
+						    &ctx->http.parser_settings,
+						    ctx->request_buf +
+						    start,
+						    len);
+			if (parsed_len <= 0) {
+				goto fail;
+			}
+
+			ctx->http.data_len = 0;
+			len = 0;
+			start = 0;
+		}
+
+		memcpy(ctx->request_buf + ctx->http.data_len,
+		       frag->data, frag->len);
+
+		ctx->http.data_len += frag->len;
+		len += frag->len;
+		frag = frag->frags;
+	}
+
+out:
+	parsed_len = http_parser_execute(&ctx->http.parser,
+					 &ctx->http.parser_settings,
+					 ctx->request_buf + start,
+					 len);
+	if (parsed_len < 0) {
+fail:
+		NET_DBG("[%p] Received %zd bytes, only parsed %d "
+			"bytes (%s %s)",
+			ctx, recv_len, parsed_len,
+			http_errno_name(ctx->http.parser.http_errno),
+			http_errno_description(
+				ctx->http.parser.http_errno));
+	}
+
+	if (ctx->http.parser.http_errno != HPE_OK) {
+		http_send_error(ctx, 400, NULL, 0);
+	} else {
+		if (ctx->state == HTTP_STATE_HEADER_RECEIVED) {
+			goto http_ready;
+		}
+
+		http_process_recv(ctx);
+	}
+
+quit:
+	http_parser_init(&ctx->http.parser, HTTP_REQUEST);
+	ctx->http.data_len = 0;
+	ctx->http.field_values_ctr = 0;
+	net_pkt_unref(pkt);
+
+	return;
+
+http_only:
+	if (ctx->cb.recv) {
+		ctx->cb.recv(ctx, pkt, 0, 0, ctx->user_data);
+	}
+
+	return;
+
+http_ready:
+	http_change_state(ctx, HTTP_STATE_OPEN);
+	url_connected(ctx, HTTP_CONNECT);
+	net_pkt_unref(pkt);
+}
+
+#if defined(CONFIG_HTTPS)
+int http_server_set_tls(struct http_ctx *ctx,
```

Comment:
A future improvement could redirect all HTTP requests to HTTPS if TLS is enabled.

Reply:
Sure, but this depends on how application is doing things. If we have enabled HTTPS, then HTTP port is not normally listened at all. So application would need to create a listener to port 80 and then re-direct traffic to HTTPS port in this case.